README
see README.md.



Changes
Revision history for Net-IP-ACL

0.01    Date/time
        First version, released on an unsuspecting world.



Makefile.PL
use 5.006;
use strict;
use warnings FATAL => 'all';
use ExtUtils::MakeMaker;

WriteMakefile(
    NAME             => 'Net::IP::ACL',
    AUTHOR           => q{Toshifumi Sakaguchi <sischkg@gmail.com>},
    VERSION_FROM     => 'lib/Net/IP/ACL.pm',
    ABSTRACT_FROM    => 'lib/Net/IP/ACL.pm',
    LICENSE          => 'MIT',
    PL_FILES         => {},
    MIN_PERL_VERSION => 5.006,
    CONFIGURE_REQUIRES => {
        'ExtUtils::MakeMaker' => 0,
        'Moose'               => 0,
        'Net::IP'             => 0,
	'Exception::Class'    => 0,
    },
    BUILD_REQUIRES => {
        'Test::More' => 0,
    },
    PREREQ_PM => {
        'ExtUtils::MakeMaker' => 0,
        'Moose'               => 0,
        'Net::IP'             => 0,
	'Exception::Class'    => 0,
    },
    dist  => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean => { FILES => 'Net-IP-ACL-*' },
);


lib/Net/IP/ACL/Bit.pm

package Net::IP::ACL::Bit;

use Moose;

has 'bit'    => ( isa      => 'Int',
		  is       => 'ro',
		  required => 1 );
has 'next_0' => ( isa     => 'Maybe[Net::IP::ACL::Bit]',
		  is      => 'rw',
		  default => undef );
has 'next_1' => ( isa => 'Maybe[Net::IP::ACL::Bit]',
		  is  => 'rw',
		  default => undef );
has 'acl_entry' => ( isa     => 'Maybe[Net::IP::ACL::Entry]',
		     is      => 'rw',
		     default => undef );

=head1 NAME

Net::IP::ACL::Bit - Node of binary tree which represents ACL.

=head1 SYNOPSIS

Quick summary of what the module does.

    use Net::IP::ACL::Bit;

    # network 01100000 00000000 00000000 00000000 = 96.0.0.0/8
    # netmask 11111111 00000000 00000000 00000000
    my $top = new Net::IP::ACL::Bit( bit = 0 );
    my $bit_0th = new Net::IP::ACL::Bit( bit => 0 ); # 0
    my $bit_1st = new Net::IP::ACL::Bit( bit => 1 ); # 1
    my $bit_2nd = new Net::IP::ACL::Bit( bit => 1 ); # 1
    my $bit_3rd = new Net::IP::ACL::Bit( bit => 0 ); # 0
    my $bit_4th = new Net::IP::ACL::Bit( bit => 0 ); # 0
    my $bit_5th = new Net::IP::ACL::Bit( bit => 0 ); # 0
    my $bit_6th = new Net::IP::ACL::Bit( bit => 0 ); # 0
    my $bit_7th = new Net::IP::ACL::Bit( bit => 0 ); # 0

    $top->next_0( $bit_0th );     # 0
    $bit_0th->next_1( $bit_1st ); # 1
    $bit_1st->next_1( $bit_2nd ); # 1
    $bit_2nd->next_0( $bit_3rd ); # 0
    $bit_3rd->next_0( $bit_4th ); # 0
    $bit_4rd->next_0( $bit_5th ); # 0
    $bit_5rd->next_0( $bit_6th ); # 0
    $bit_6rd->next_0( $bit_7th ); # 0


=head1 SUBROUTINES/METHODS

=head2

=head2 new( bit => $bit )

=head2 next_0( $bit )

set 0-node child.

=head2 next_0()

get 0-node child.

=head2 next_1( $bit )

set 1-node child.

=head2 next_1()

get 1-node child.

=head2 print_acl_bit()

print debug message to STDERR.

=cut

sub print_acl_bit {
    my $this = shift;

    printf STDERR "bit %d, next_1: %d, next_0: %d, acl_entry: %s\n",
	$this->bit, $this->next_1 ? 1 : 0, $this->next_0 ? 1 : 0, $this->acl_entry ? $this->acl_entry->name : "none";
}

no Moose;
__PACKAGE__->meta->make_immutable;

1;



lib/Net/IP/ACL/Imp.pm

package Net::IP::ACL::Imp;

use Moose;
use Net::IP::ACL::Bit;
use Net::IP::ACL::Entry;
use Data::Dumper;

has '_top' => ( isa     => 'Net::IP::ACL::Bit',
		is      => 'ro',
		default => sub { new Net::IP::ACL::Bit( bit => 0 ) } );
has '_bit_length' => ( isa      => 'Int',
		       is       => 'ro',
		       required => 1 );

=head1 NAME

Net::IP::ACL::Imp - Implementation of Access Control List.

=head1 SYNOPSIS

Quick summary of what the module does.

    use Net::IP::ACL::Imp;

    my $my_network = new Net::IP::ACL::Entry(
        address => '172.16.10.0/24',
        name    => 'my network',
        value   => 1,
    );

    # make ACL for IPv4.
    my $_acl_v4 = new Net::IP::ACL::Imp( _bit_length => 32 );
    $_acl_v4->add( $my_network );
    $_acl_v4->add( $group_network );
    $_acl_v4->add( $my_host );

    $_acl_v4->match( '10.0.0.1' );                  # unmatched, return undef.

    $_acl_v4->match( '172.16.10.100' );             # matched, return $.

    print $acl_entry3->name;                        # "my host"

=head1 SUBROUTINES/METHODS

=head2 new( _bit_length => $bit_length )

Bit length of IP address, Bit length of IPv4 address length is 32, bit length of IPv6 address is 128.

=head2 add( $acl_entry )

add acl_entry, and return self.

=cut

sub add {
    my $this = shift;
    my ( $ac ) = @_;

    my $node = $this->_top();
    for ( my $i = 1 ; $i <= $this->_bit_length + 1; $i++ ) {
	my $bit = Net::IP::ACL::Entry::bit_n( $this->_bit_length - $i );
	if ( ! ( $ac->netmask & $bit ) ) {
	    $node->acl_entry( $ac );
	    last;
	}

	if ( $ac->network & $bit ) {
	    if ( ! defined( $node->next_1 ) ) {
		$node->next_1( new Net::IP::ACL::Bit( bit => 1 ) );
	    }
	    $node = $node->next_1;
	}
	else {
	    if ( ! defined( $node->next_0 ) ) {
		$node->next_0( new Net::IP::ACL::Bit( bit => 0 ) );
	    }
	    $node = $node->next_0;
	}
    }

    return $this;
}


=head2 match( $address )

whether IP address $address is matched or not.
If $address is matched, this method return matched Net::IP::ACL::Entry.
If not matched, this method return undef.

=cut

sub match {
    my $this = shift;
    my ( $address ) = @_;

    my $address_binary    = $address->intip();
    my $node              = $this->_top();
    my $matched_acl_entry = undef;
    my $i                 = 0;

    while ( $node ) {
	my $bit = Net::IP::ACL::Entry::bit_n( $this->_bit_length - 1 - $i );

	if ( $node->acl_entry ) {
	    $matched_acl_entry = $node->acl_entry;
	}
	if ( $bit & $address_binary ) {
	    $node = $node->next_1;
	}
	else {
	    $node = $node->next_0;
	}
	$i++;
    }

    return $matched_acl_entry;
}


=head2 print_bits

Print bit pattern to STDERR for debug.

=cut

sub print_bits {
    my $this = shift;
    my ( $num ) = @_;

    for ( my $i = 0 ; $i < $this->_bit_length ; $i++ ) {
	if ( $i % 8 == 0 ) {
	    print STDERR " ";
	}
	my $bit = Net::IP::ACL::Entry::bit_n( $this->_bit_length - $i - 1 );
	printf STDERR "%d", $bit & $num ? 1 : 0;
    }
    print STDERR "\n";
}

1;


lib/Net/IP/ACL/Entry.pm
package Net::IP::ACL::Entry;

use Moose;
use Net::IP;
use Net::IP::ACL::Exception;
use Math::BigInt;
use Data::Dumper;

has 'address' => ( isa => 'Net::IP',    is => 'ro', required => 1 );
has 'name'    => ( isa => 'Maybe[Str]', is => 'ro' );
has 'value'   => ( isa => 'Any',        is => 'rw' );

around BUILDARGS => sub {
    my $orig  = shift;
    my $class = shift;
    my $args  = $class->$orig( @_ );

    if ( ! exists( $args->{address} ) ) {
	Net::IP::ACL::ArugmentError->throw(
	    error_message => "new Net::IP::ACL::Entry must be specified address.",
	);
    }

    if ( ! defined( blessed( $args->{address} ) ) ) {
	$args->{address} = new Net::IP( $args->{address} );
	if ( ! $args->{address} ) {
	    Net::IP::ACL::ArugmentError->throw(
		error_message => q{address must be Net::IP instance or "network/netmask" string},
	    );
	}
    }

    return $args;
};


=head1 NAME

Net::IP::ACL::Entry - Network ACL entry.

=head1 SYNOPSIS

Quick summary of what the module does.

    # log server
    use Net::IP::ACL::Entry;

    my $my_network = new Net::IP::ACL::Entry(
        address => '172.16.10.0/24',
        name    => 'my network',
        value   => 1,
    );


=head1 EXPORT

=head1 SUBROUTINES/METHODS

=head2 network

return IPv4 or IPv6 address by binary form.

=cut

sub network {
    my $this = shift;
    return new Math::BigInt( $this->address()->intip() );
}

=head2 netmask

return IPv4 or IPv6 netmask by binary form.

=cut

sub netmask {
    my $this = shift;

    my $bit_length = $this->version() == 4 ? 32 : 128;
    return bit_n( $bit_length ) - bit_n( $bit_length - $this->address()->prefixlen() );
}


=head2 bit_n( $n )

return 0x01 << $n (bit shift operator) by Math::BigInt. if $n < 0 return 0.

=cut

sub bit_n {
    my ( $n ) = @_;

    if ( $n >= 0 ) {
	return Math::BigInt->new( 1 )->blsft( $n );
    }
    else {
	return Math::BigInt->new( 0 );
    }
}

=head2 version

If network address is IPv4, return 4. If network address is IPv6, return 6.

=cut

sub version {
    my $this = shift;

    return $this->address->version();
}

=head2 check_ip_address( $str )

Check $str is IP address format string, and return Net::IP instance.
If $str is not IP address format, throw ArgumentError exception.

=cut

sub check_ip_address {
    my ( $str ) = @_;

    my $ip_addr = new Net::IP( $str );
    if ( ! defined( $ip_addr ) ) {
	Net::IP::ACL::ArgumentError->throw(
	    error_message => sprintf( q{Invalid network address "%s"(%s).}, $str, Net::IP::Error() ),
	);
    }
    return $ip_addr;
}


no Moose;
__PACKAGE__->meta->make_immutable;

1;


lib/Net/IP/ACL.pm
package Net::IP::ACL;

use Moose;
use Net::IP;
use Net::IP::ACL::Exception;
use Net::IP::ACL::Entry;
use Net::IP::ACL::Imp;


has '_ipv4_acl' => ( isa     => 'Net::IP::ACL::Imp',
		     is      => 'rw',
		     default => sub { new Net::IP::ACL::Imp( _bit_length => 32 ) } );

has '_ipv6_acl' => ( isa     => 'Net::IP::ACL::Imp',
		     is      => 'rw',
		     default => sub { new Net::IP::ACL::Imp( _bit_length => 128 ) } );

=head1 NAME

Net::IP::ACL - ACL.

=head1 VERSION

=cut

our $VERSION = '0.1.1';

=head1 SYNOPSIS

Quick summary of what the module does.

    use Net::IP::ACL;

    my $my_network = new Net::IP::ACL::Entry(
        address => '172.16.10.0/24',
        name    => 'my network',
        value   => 1,
    );

    my $group_network = new Net::IP::ACL::Entry(
        address    => '172.16.10.0/16',
        name       => 'group network',
        value      => 2,
    );

    my $my_host = new Net::IP::ACL::Entry(
        address    => '172.16.10.101',
        name       => 'my host',
        value      => 0,
    );

    # make ACL.
    my $acl = new Net::IP::ACL;
    $acl->add( $my_network );
    $acl->add( $group_network );
    $acl->add( $my_host );

    $acl->match( '10.0.0.1' );                      # unmatched, return undef.

    my $acl_entry1 = $acl->match( '172.16.230.24' ) # matched $group_network, return $group_network.
    print $acl_entry1->name;                        # "group network"

    my $acl_entry2 = $acl->match( '172.16.10.24' )  # matched $my_network, return $my_network.
    print $acl_entry2->name;                        # "my network"

    my $acl_entry3 = $acl->match( '172.16.10.101' ) # matched $my_host, return $my_host.
    print $acl_entry3->name;                        # "my host"


=head1 SUBROUTINES/METHODS

=head2

=head2 add( $acl_entry )

add acl_entry.

=cut

sub add {
    my $this = shift;
    my ( $acl_entry ) = @_;

    my $acl = $this->_get_acl( $acl_entry->address() );
    $acl->add( $acl_entry );
}

=head2 match( $address )

whether IP address $address is matched or not.
If $address is matched, this method return matched ACLEntry.
If not matched, this method return undef.

=cut

sub match {
    my $this = shift;
    my ( $str ) = @_;

    my $address = Net::IP::ACL::Entry::check_ip_address( $str );
    my $acl     = $this->_get_acl( $address );
    return $acl->match( $address );
}


sub _get_acl {
    my $this = shift;
    my ( $address ) = @_;

    if ( $address->version() == 4 ) {
	return $this->_ipv4_acl;
    }
    elsif ( $address->version() == 6 ) {
	return $this->_ipv6_acl;
    }
    else {
	Net::IP::ACL::ArgumentError->throw(
	    error_message => sprintf( q{unknown protocol version "%d"}, $address->version ),
	);
    }
}

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Net::IP::ACL

=head1 LICENSE AND COPYRIGHT

Copyright 2015 Toshifumi Sakaguchi.

This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (2.0). You may obtain a
copy of the full license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any use, modification, and distribution of the Standard or Modified
Versions is governed by this Artistic License. By using, modifying or
distributing the Package, you accept this license. Do not use, modify,
or distribute the Package, if you do not accept this license.

If your Modified Version has been derived from a Modified Version made
by someone other than you, you are nevertheless required to ensure that
your Modified Version complies with the requirements of this license.

This license does not grant you the right to use any trademark, service
mark, tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge
patent license to make, have made, use, offer to sell, sell, import and
otherwise transfer the Package with respect to any patent claims
licensable by the Copyright Holder that are necessarily infringed by the
Package. If you institute patent litigation (including a cross-claim or
counterclaim) against any party alleging that the Package constitutes
direct or contributory patent infringement, then this Artistic License
to you shall terminate on the date that such litigation is filed.

Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

=cut

1; # End of Net::IP::ACL

